OneSignal/OneSignal-iOS-SDK
# Live activity support 5.0.0

# Description
## One Line Summary
Add Live Activity support to Major release 5.0.0

## Details
Live Activities are a type of interactive push notification. They were designed by Apple to enable iOS apps to provide real-time updates to their users that are visible from the lock screen and dynamic island.

```
let activity = try Activity<OneSignalWidgetAttributes>.request(
  attributes: attributes,
  contentState: contentState,
  pushType: .token)
Task {
    for await data in activity.pushTokenUpdates {
        let myToken = data.map {String(format: "%02x", $0)}.joined()
        // ... required code for entering a live activity
        OneSignal.enterLiveActivity("my_activity_id", withToken: myToken)
    }
}
```
## Functional Testing instructions
Basic case
1. Prepare to make a Live Activity Update

< app_id > : Found on web console
< authorization token for app_id > : Found on web console
< activity_id > : Typed into app at runtime
```
curl --request POST \
     --url https://onesignal.com/api/v1/apps/< app_id >/live_activities/< activity_id >/notifications \
     --header 'Authorization: Basic < authorization token for app_id >' \
     --header 'accept: application/json' \
     --header 'content-type: application/json' \
     --data '
{
     "event": "update",
     "event_updates": {
          "message" : "Hello from OneSignal!"
     },
     "name": "test update"
}
'

```

2. Run the app
3. Scroll to the bottom
3. Type in the chosen activity_id
4. Touch enter live activity
5. Pull down from the upper left
6. See the live activity
7. Run the CURL command above
8. See update!

Case 2: Delayed subscription id
in this case the subscription_id is not available so we need to cache the enter live activity request

in this case we remove
```
+ (void) initialize {
 subscriptionId = OneSignalUserManagerImpl.sharedInstance.pushSubscriptionId;
           
}
```
and add
```
dispatch_async(dispatch_get_main_queue(), ^{
        [NSTimer scheduledTimerWithTimeInterval:30 repeats:NO block:^(NSTimer *timer) {
            subscriptionId = OneSignalUserManagerImpl.sharedInstance.pushSubscriptionId;
            [OneSignalLiveActivityController executePendingLiveActivityUpdates];
        }];
       });
```
at the engining of enterLiveActivity inside the `OneSignalLiveActivityController


 1. Start a live activity
 2. Quickly make an update and see that it does not go through
 3. Wait 30 seconds. (add a breakpoint to the timer block if you want to be sure)
 4. Make an update and see that it now goes through.

## Unit testing
Added several tests to test the requests and the public SDK method

## Manual testing
I have tested Live Activity updates and they work properly.

# Affected code checklist

# Checklist
## Overview
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR

## Testing

## Final pass
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1189)
